lib/msal-common/src/server/ServerTokenRequestParameters.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { ICrypto } from "../crypto/ICrypto";
+import { ScopeSet } from "../auth/ScopeSet";
+import { TokenExchangeParameters } from "../request/TokenExchangeParameters";
+import { ProtocolUtils } from "../utils/ProtocolUtils";
+import { Constants, HEADER_NAMES, AADServerParamKeys } from "../utils/Constants";
+import { ServerRequestParameters } from "./ServerRequestParameters";
+import { CodeResponse } from "../response/CodeResponse";
+
+export class ServerTokenRequestParameters extends ServerRequestParameters {
+
+    // Params
+    tokenRequest: TokenExchangeParameters;
+    codeResponse: CodeResponse;
+
+    constructor(clientId: string, tokenRequest: TokenExchangeParameters, codeResponse: CodeResponse, redirectUri: string, cryptoImpl: ICrypto) {
+        super(clientId, redirectUri, cryptoImpl);
+        this.tokenRequest = tokenRequest;
+        this.codeResponse = codeResponse;
+
+        this.scopes = new ScopeSet(this.tokenRequest && this.tokenRequest.scopes, this.clientId, false);
+
+        const randomGuid = this.cryptoObj.createNewGuid();
+        this.state = ProtocolUtils.setRequestState(this.codeResponse && this.codeResponse.userRequestState, randomGuid);
+
+        this.correlationId = this.tokenRequest.correlationId || this.cryptoObj.createNewGuid();
+    }
+
+    createRequestHeaders(): Map<string, string> {
+        const headers = new Map<string, string>();
+        headers.set(HEADER_NAMES.CONTENT_TYPE, Constants.URL_FORM_CONTENT_TYPE);
+        return headers;
+    }
+
+    createRequestBody(): string {
+        const paramString = this.createParamString();
+        return paramString.join("&");
+    }
+
+    protected createParamString(): Array<string> {
+        const str: Array<string> = [];
+        this.replaceDefaultScopes();
```

Comment:
I can add this as a function to ScopeSet, but I don't think we want to have it happen immediately, since we want to track whether client id was sent or not.

Reply:
changed